tools/yulPhaser/Population.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <tools/yulPhaser/Population.h>
+
+#include <tools/yulPhaser/Program.h>
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <numeric>
+
+using namespace std;
+using namespace solidity::langutil;
+
+namespace solidity::phaser
+{
+
+size_t Population::measureFitness(Chromosome const& _chromosome, CharStream& _sourceCode)
```

Comment:
I don't see much sense in having this static if you pass almost half the state anyway :)

Reply:
Yeah, this is not the best design but in this PR the `Population` class is still just a stopgap solution to have the algoritm do something.

The mutation/crossover/evaluation part will change completely. Fitness metric will be a separate base class with various specializations (e.g. code size). It will be passed to population's constructor. Similarly, I'll have base classes for mutations, crossover and selections and specializations. This will allow me to use them as building blocks to easily construct and test different variations of the algorithm in later phases (in this phase I'm just doing the base classes and one or two specializations for each). I'm also going to make evaluation implicit (store chromosomes in an ordered set, always evaluated and sorted and make population evaluate new chromosomes as soon as they are created/added/mutated).

Below is an example of how an algorithm will look like using this approach.

``` c++
// One round of the algorithm, inside GeneticAlgorithm::run()

Population populationWithMutations = population;
populationWithMutations.mutateInPlace(RandomSelection(), GeneRandomization(RandomGenes(0.1)));
populationWithMutations.mutateInPlace(RandomSelection(), GeneDeletion(RandomGenes(0.1)));
Population recombinations = populationWithMutations.crossover(RandomPairSelection(), HalfSwap(0.5));

return (populationWithMutations + recombinations).select(TopSelection());
```

This would be initialized with:
``` c++
auto sourceCode = make_shared<Program>(program);
GeneticAlgorithm algorithm(ProgramSize(sourceCode), cout);
algorithm.run();
```

But that's a big change in the upcoming PR. In this one I'd prefer to just leave it as is.